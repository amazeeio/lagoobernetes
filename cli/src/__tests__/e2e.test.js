// @flow

import fs from 'fs';
import path from 'path';
import R from 'ramda';
import { sync as spawnSync } from 'execa';

const CLI_PATH = path.join(__dirname, '..', '..', 'bin', 'lagu.js');
const cwd = path.join(__dirname, 'fixtures');
const tokenPath = path.join(cwd, '.lagoobernetes-token');

const stripCreatedDates = R.replace(
  /\d{4}-\d{2}-\d{2} \d{2}:\d{2}:\d{2}/g,
  '                   ',
);

describe('lagu', () => {
  it('should fail with error message without any arguments', async () => {
    const results = spawnSync(CLI_PATH, [], {
      cwd,
      reject: false,
    });
    expect(results.code).toBe(1);
    expect(results.message).toMatch('Not enough non-option arguments');
  });

  it('should init', async () => {
    const results = spawnSync(
      CLI_PATH,
      [
        'init',
        '--overwrite',
        '--token',
        tokenPath,
        '--project',
        'ci-github',
        '--api',
        'http://localhost:3000',
        '--ssh',
        'localhost:2020',
      ],
      {
        cwd,
      },
    );
    expect(results.code).toBe(0);
    expect(results.stdout).toMatchSnapshot();
  });

  it('should not error on logout when not logged in', async () => {
    const results = spawnSync(CLI_PATH, ['logout'], {
      cwd,
    });
    expect(results.code).toBe(0);
    expect(results.stdout).toMatchSnapshot();
  });

  it('should log in', async () => {
    const results = spawnSync(
      CLI_PATH,
      [
        'login',
        '--identity',
        path.join('..', '..', '..', '..', 'local-dev', 'cli_id_rsa'),
      ],
      {
        cwd,
      },
    );
    expect(results.code).toBe(0);
    expect(results.stdout).toMatchSnapshot();
    // Test whether token file has JWT header
    const tokenHeader = R.compose(
      R.nth(0),
      R.split('.'),
    )(fs.readFileSync(tokenPath, 'utf8'));
    expect(tokenHeader).toMatch(/^eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9$/);
  });

  it('should show customer details (using --project option)', async () => {
    const results = spawnSync(
      CLI_PATH,
      ['customer', '--project', 'ci-multiproject1'],
      {
        cwd,
      },
    );
    expect(results.code).toBe(0);
    expect(stripCreatedDates(results.stdout)).toMatchSnapshot();
  });

  it('should show customer details (project read from .lagoobernetes.yml)', async () => {
    const results = spawnSync(CLI_PATH, ['customer'], {
      cwd,
    });
    expect(results.code).toBe(0);
    expect(stripCreatedDates(results.stdout)).toMatchSnapshot();
  });

  it('should list all environments for a project (using --project option)', async () => {
    const results = spawnSync(
      CLI_PATH,
      ['environments', '--project', 'ci-github'],
      {
        cwd,
      },
    );
    expect(results.code).toBe(0);
    expect(stripCreatedDates(results.stdout)).toMatchSnapshot();
  });

  it('should list no environments for a project with none (using --project option)', async () => {
    const results = spawnSync(
      CLI_PATH,
      ['environments', '--project', 'ci-multiproject1'],
      {
        cwd,
      },
    );
    expect(results.code).toBe(0);
    expect(stripCreatedDates(results.stdout)).toMatchSnapshot();
  });

  it('should log an error for a non-existent project (using --project option)', async () => {
    const results = spawnSync(
      CLI_PATH,
      ['environments', '--project', 'non-existent-project'],
      {
        cwd,
      },
    );
    expect(results.code).toBe(0);
    expect(stripCreatedDates(results.stdout)).toMatchSnapshot();
  });

  it('should list all environments for a project (project read from .lagoobernetes.yml)', async () => {
    const results = spawnSync(CLI_PATH, ['environments'], { cwd });
    expect(results.code).toBe(0);
    expect(stripCreatedDates(results.stdout)).toMatchSnapshot();
  });

  it('should show project details (read from .lagoobernetes.yml)', async () => {
    const results = spawnSync(CLI_PATH, ['project'], {
      cwd,
    });
    expect(results.code).toBe(0);
    expect(stripCreatedDates(results.stdout)).toMatchSnapshot();
  });

  it('should list all projects', async () => {
    const results = spawnSync(CLI_PATH, ['projects'], {
      cwd,
    });
    expect(results.code).toBe(0);
    expect(stripCreatedDates(results.stdout)).toMatchSnapshot();
  });

  it('should create a project', async () => {
    const results = spawnSync(
      CLI_PATH,
      [
        'project',
        'create',
        '--customer',
        '3',
        '--name',
        'e2e-test-project',
        '--gitUrl',
        'ssh://git@172.17.0.1:2222/git/e2e-test-project.git',
        '--openshift',
        '2',
        '--branches',
        'true',
        '--pullrequests',
        'true',
        '--productionEnvironment',
        'master',
      ],
      {
        cwd,
      },
    );
    expect(results.code).toBe(0);
    expect(stripCreatedDates(results.stdout)).toMatchSnapshot();
  });

  it('should show newly-created project details (using --project option)', async () => {
    const results = spawnSync(
      CLI_PATH,
      ['project', '--project', 'e2e-test-project'],
      {
        cwd,
      },
    );
    expect(results.code).toBe(0);
    expect(stripCreatedDates(results.stdout)).toMatchSnapshot();
  });

  it('should list the new project among all projects', async () => {
    const results = spawnSync(CLI_PATH, ['projects'], {
      cwd,
    });
    expect(results.code).toBe(0);
    expect(stripCreatedDates(results.stdout)).toMatchSnapshot();
  });

  it('should delete the project', async () => {
    const results = spawnSync(
      CLI_PATH,
      ['project', 'delete', '--project', 'e2e-test-project'],
      {
        cwd,
      },
    );
    expect(results.code).toBe(0);
    expect(results.stdout).toMatchSnapshot();
  });

  it('should not show deleted project details (using --project option)', async () => {
    const results = spawnSync(
      CLI_PATH,
      ['project', '--project', 'e2e-test-project'],
      {
        cwd,
      },
    );
    expect(results.code).toBe(0);
    expect(stripCreatedDates(results.stdout)).toMatchSnapshot();
  });

  it('should log out when logged in', async () => {
    const results = spawnSync(CLI_PATH, ['logout'], {
      cwd,
    });
    expect(results.code).toBe(0);
    expect(results.stdout).toMatchSnapshot();
  });
});
