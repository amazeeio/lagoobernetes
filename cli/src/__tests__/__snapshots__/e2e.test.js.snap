// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lagu should create a project 1`] = `
"! Set \\"customer\\" option to \\"3\\" from arguments or config
! Set \\"name\\" option to \\"e2e-test-project\\" from arguments or config
! Set \\"gitUrl\\" option to \\"ssh://git@172.17.0.1:2222/git/e2e-test-project.git\\" from arguments or config
! Set \\"openshift\\" option to \\"2\\" from arguments or config
! Set \\"branches\\" option to \\"true\\" from arguments or config
! Set \\"pullrequests\\" option to \\"true\\" from arguments or config
! Set \\"productionEnvironment\\" option to \\"master\\" from arguments or config
Project \\"e2e-test-project\\" created successfully:
Name                   e2e-test-project
Customer               ci-customer
Git URL                ssh://git@172.17.0.1:2222/git/e2e-test-project.git
Active Systems Deploy  lagoobernetes_openshiftBuildDeploy
Active Systems Remove  lagoobernetes_openshiftRemove
Branches               true
Pull Requests          true
Openshift              ci-local
Created                                   "
`;

exports[`lagu should delete the project 1`] = `"Project \\"e2e-test-project\\" deleted successfully!"`;

exports[`lagu should init 1`] = `
"! Set \\"overwrite\\" option to \\"true\\" from arguments or config
! Set \\"project\\" option to \\"ci-github\\" from arguments or config
! Set \\"api\\" option to \\"http://localhost:3000\\" from arguments or config
! Set \\"ssh\\" option to \\"localhost:2020\\" from arguments or config
! Set \\"token\\" option to \\"/Users/k/p/lagoobernetes/cli/src/__tests__/fixtures/.lagoobernetes-token\\" from arguments or config
Creating file '/Users/k/p/lagoobernetes/cli/src/__tests__/fixtures/.lagoobernetes.yml'...
Configuration file created!"
`;

exports[`lagu should list all environments for a project (project read from .lagoobernetes.yml) 1`] = `
"Name              master
Environment Type  production
Deploy Type       branch
Created
Updated                              "
`;

exports[`lagu should list all environments for a project (using --project option) 1`] = `
"Name              master
Environment Type  production
Deploy Type       branch
Created
Updated                              "
`;

exports[`lagu should list all projects 1`] = `
"Project                Git URL                                                 Branches  Pull Requests  Created
ci-bitbucket           ssh://git@172.17.0.1:2222/git/bitbucket.git           true      true
ci-drupal              ssh://git@172.17.0.1:2222/git/drupal.git              true      true
ci-drupal-galera       ssh://git@172.17.0.1:2222/git/drupal-galera.git       true      true
ci-drupal-postgres     ssh://git@172.17.0.1:2222/git/drupal-postgres.git     true      true
ci-elasticsearch       ssh://git@172.17.0.1:2222/git/elasticsearch.git       true      true
ci-env-limit           ssh://git@172.17.0.1:2222/git/node.git                true      true
ci-features            ssh://git@172.17.0.1:2222/git/features.git            true      true
ci-features-subfolder  ssh://git@172.17.0.1:2222/git/features-subfolder.git  true      true
ci-github              ssh://git@172.17.0.1:2222/git/github.git              true      true
ci-gitlab              ssh://git@172.17.0.1:2222/git/gitlab.git              true      true
ci-multiproject1       ssh://git@172.17.0.1:2222/git/multiproject.git        true      true
ci-multiproject2       ssh://git@172.17.0.1:2222/git/multiproject.git        true      true
ci-nginx               ssh://git@172.17.0.1:2222/git/nginx.git               true      true
ci-node                ssh://git@172.17.0.1:2222/git/node.git                true      true
ci-rest                ssh://git@172.17.0.1:2222/git/rest.git                true      true                              "
`;

exports[`lagu should list no environments for a project with none (using --project option) 1`] = `"Name  Environment Type  Deploy Type  Created  Updated"`;

exports[`lagu should list the new project among all projects 1`] = `
"Project                Git URL                                                 Branches  Pull Requests  Created
ci-bitbucket           ssh://git@172.17.0.1:2222/git/bitbucket.git           true      true
ci-drupal              ssh://git@172.17.0.1:2222/git/drupal.git              true      true
ci-drupal-galera       ssh://git@172.17.0.1:2222/git/drupal-galera.git       true      true
ci-drupal-postgres     ssh://git@172.17.0.1:2222/git/drupal-postgres.git     true      true
ci-elasticsearch       ssh://git@172.17.0.1:2222/git/elasticsearch.git       true      true
ci-env-limit           ssh://git@172.17.0.1:2222/git/node.git                true      true
ci-features            ssh://git@172.17.0.1:2222/git/features.git            true      true
ci-features-subfolder  ssh://git@172.17.0.1:2222/git/features-subfolder.git  true      true
ci-github              ssh://git@172.17.0.1:2222/git/github.git              true      true
ci-gitlab              ssh://git@172.17.0.1:2222/git/gitlab.git              true      true
ci-multiproject1       ssh://git@172.17.0.1:2222/git/multiproject.git        true      true
ci-multiproject2       ssh://git@172.17.0.1:2222/git/multiproject.git        true      true
ci-nginx               ssh://git@172.17.0.1:2222/git/nginx.git               true      true
ci-node                ssh://git@172.17.0.1:2222/git/node.git                true      true
ci-rest                ssh://git@172.17.0.1:2222/git/rest.git                true      true
e2e-test-project       ssh://git@172.17.0.1:2222/git/e2e-test-project.git    true      true                              "
`;

exports[`lagu should log an error for a non-existent project (using --project option) 1`] = `"No project 'non-existent-project' found."`;

exports[`lagu should log in 1`] = `
"Logging in to lagoobernetes at lagoobernetes@localhost:2020...
Logged in successfully."
`;

exports[`lagu should log out when logged in 1`] = `"Logged out successfully."`;

exports[`lagu should not error on logout when not logged in 1`] = `"Logged out successfully."`;

exports[`lagu should not show deleted project details (using --project option) 1`] = `"No project 'e2e-test-project' found."`;

exports[`lagu should show customer details (project read from .lagoobernetes.yml) 1`] = `
"Customer details for project 'ci-github':
Name                ci-customer
Comment             null
Deploy Private Key  √
Users               4, 5
Created                                "
`;

exports[`lagu should show customer details (using --project option) 1`] = `
"Customer details for project 'ci-multiproject1':
Name                ci-customer
Comment             null
Deploy Private Key  √
Users               4, 5
Created                                "
`;

exports[`lagu should show newly-created project details (using --project option) 1`] = `
"Name                   e2e-test-project
Customer               ci-customer
Git URL                ssh://git@172.17.0.1:2222/git/e2e-test-project.git
Active Systems Deploy  lagoobernetes_openshiftBuildDeploy
Active Systems Remove  lagoobernetes_openshiftRemove
Branches               true
Pull Requests          true
Openshift              ci-local
Created                                   "
`;

exports[`lagu should show project details (read from .lagoobernetes.yml) 1`] = `
"Name                   ci-github
Customer               ci-customer
Git URL                ssh://git@172.17.0.1:2222/git/github.git
Active Systems Deploy  lagoobernetes_openshiftBuildDeploy
Active Systems Remove  lagoobernetes_openshiftRemove
Branches               true
Pull Requests          true
Openshift              ci-local
Created                                   "
`;
